File: DATA/Boxscores/LIBRARY/BoxscoresClass.py
import sqlite3
import pandas as pd

class Boxscores:
  
    
    def __init__(self, path_to_database=""):
        
        self.path = path_to_database
    
    def boxscores_dataframe(self):
       
        # Connect to database
        con = sqlite3.connect(self.path)
        
        try:
            # Read tables
            boxscores = pd.read_sql_query("SELECT * FROM Boxscores", con)
            players = pd.read_sql_query("SELECT * FROM Players", con)
            teams = pd.read_sql_query("SELECT * FROM Teams", con)
            
            # Perform left joins
            results_df = boxscores.merge(players, on='PLAYER_ID', how='left')
            results_df = results_df.merge(teams, on='TEAM_ID', how='left')
            
            # Drop PLAYER_ID and TEAM_ID columns
            results_df = results_df.drop(columns=['PLAYER_ID', 'TEAM_ID'])
            
            return results_df
            
        finally:
            # Always close the connection
            con.close()


def get_boxscores_instance(db_path):

    return Boxscores(path_to_database=db_path)

File: DATA/Boxscores/LIBRARY/BoxscoresClass.R
library(RSQLite)
library(DBI)
library(S7)
library(dplyr)


Boxscores <- new_class(
  "Boxscores",
  properties = list(
    path = new_property(class_character)
  ),
  # Custom constructor to set the path property
  constructor = function(path_to_database = character()) {
    new_object(S7_object(), path = path_to_database)
  }
)

# Create a new generic function for retrieving the data frame
boxscores_dataframe <- new_generic("Boxscores_dataframe", "x")

method(boxscores_dataframe, Boxscores) <- function(x) {

  con <- dbConnect(RSQLite::SQLite(), x@path)
  
  query <- tbl(con, "Boxscores") |>
      left_join(tbl(con, "Players"), by = "PLAYER_ID") |>
        left_join(tbl(con, "Teams"), by = "TEAM_ID") |>
          select(-PLAYER_ID, -TEAM_ID)
    
  results_df <- query |>
    collect()
    
    dbDisconnect(con)
    
    return(results_df)
}

get_Boxscores_instance <- function(db_path) {
  Boxscores(path_to_database = db_path)
}

get_Boxscores_data <- function() {
  path_to_database <- paste0(Sys.getenv("EXAMPLES"), "Boxscores.db")
  boxscores_object <- get_Boxscores_instance(db_path = path_to_database)
  boxscores_data   <-  boxscores_dataframe(boxscores_object)
  return(boxscores_data)
}
